### PR TITLE
fix: stop pre-allocating and undercounting a hash table per COUNT(DISTINCT) group

### PR DESCRIPTION
**Pull request by adriangb**
A grouped `COUNT(DISTINCT <string>)` over 4,000 groups holding 2 short strings each needs a 36 MB memory budget. It needs 2.0 MB after this change.

Every group gets its own hash table, and each table is allocated at warm-up size before the group holds anything, so the memory the query needs tracks the number of groups rather than the amount of data. The query also reports less memory than it holds, so a memory limit does not stop it at the right point.

## Reproduction

This needs only `datafusion-cli`. There is no patch, no custom allocator and no data file.

```sql
-- repro.sql
SET datafusion.execution.target_partitions = 1;

-- 4,000 groups with 2 distinct short strings in each.
-- avg() stops SingleDistinctToGroupBy from rewriting the distinct aggregate away.
SELECT g, count(DISTINCT s) AS d, avg(p) AS a
FROM (
  SELECT v % 4000 AS g, 'v' || CAST(v AS VARCHAR) AS s, v AS p
  FROM generate_series(0, 7999) AS t(v)
)
GROUP BY g
ORDER BY g
LIMIT 3;
```

```
datafusion-cli -m 8M -f repro.sql
```

`s` is a `Utf8View` column, so this exercises `ArrowBytesViewMap`. The 8,000 rows arrive in one batch, so the aggregate builds all 4,000 accumulators before it can emit or spill.

Current `main` at `20d1c56761` fails:

```
Resources exhausted: Additional allocation failed for SingleHashAggregateStream[0] with top memory
consumers (across reservations) as:
  DataFusion-Cli#1(can spill: false) consumed 0.0 B, peak 0.0 B,
  SingleHashAggregateStream[0]#2(can spill: true) consumed 0.0 B, peak 48.0 B,
  TopK[0]#3(can spill: false) consumed 0.0 B, peak 0.0 B.
Error: Failed to allocate additional 111.1 MB for SingleHashAggregateStream[0] with 0.0 B already
allocated for this reservation - 8.0 MB remain available for the total memory pool:
greedy(used: 0.0 B, pool_size: 8.0 MB)
```

This branch returns the rows:

```
+---+---+--------+
| g | d | a      |
+---+---+--------+
| 0 | 2 | 2000.0 |
| 1 | 2 | 2001.0 |
| 2 | 2 | 2002.0 |
+---+---+--------+
3 row(s) fetched.
Elapsed 0.006 seconds.
```

Those 4,000 accumulators hold 8,000 short strings, which is about 100 KB of data. The base asks the pool for 111.1 MB to hold it. This branch runs the same query inside `-m 3M`. Both builds return the same rows, and the base does so at `-m 200M`. Each run takes well under a second, and the outcome repeats exactly over three runs on each side.

## How much it improves

The minimum memory limit at which that query completes, bisected on each side:

| value column | before | after |
| --- | --- | --- |
| `Utf8` | fails 34 MB, passes 36 MB | fails 1.8 MB, passes 2.0 MB |
| `Utf8View` | fails 120 MB, passes 124 MB | fails 2.4 MB, passes 2.6 MB |

`clickbench_extended` at `DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G`, pool peak over six runs:

| query | base | this branch | change |
| --- | --- | --- | --- |
| Q2, grouped, 4 string distincts | 98.1 to 98.6 MiB | 11.7 MiB in all six runs | -88.1% |
| Q1, ungrouped string distincts | 3.4 MiB | 2.4 MiB | -29.4% |
| Q0, ungrouped, high cardinality | 796.8 to 834.8 MiB | 846.5 to 885.5 MiB | +6.3% |

Q2 is the only query in any benchmark suite that puts a grouped `COUNT(DISTINCT)` on a non-integer column.

Q0 costs more, and it is the one disclosed cost of this PR. Those extra bytes are memory Q0 always held and the pool could not see, not new allocation; appendix A has the three-build decomposition that separates the two. Latency does not move anywhere, which is what an allocation-sizing change should do.

## Which issue does this PR close?

No existing issue. I found this when I investigated a production out of memory. I can file an issue if you want it in the changelog.

## Rationale for this change

Pre-allocating is right for the one long-lived map behind a `GROUP BY` on a string column. It is wrong for the distinct-count accumulators, because `GroupsAccumulatorAdapter` creates one accumulator per group and most groups hold a handful of values. There the warm-up dwarfs the data.

Both maps also under-report the table they hold. `ArrowBytesViewMap` left the control bytes out. `ArrowBytesMap` charged the table only when it grew, so a map that stayed inside its pre-allocation reported its table as free forever. A memory limit acted on a number that was too small.

`clear_shrink` is the third part. The aggregate stream calls it to hand memory back before it spills and before a downstream sort. It restored the warm-up capacity instead of releasing it, so nothing came back.

## What changes are included in this PR?

- `new` on both maps allocates nothing. A new `with_capacity` keeps the previous pre-allocating behavior. `GroupValuesBytes` and `GroupValuesBytesView` use `with_capacity`, and the two distinct-count accumulators use `new`. A map remembers how it was built, so `take` warms it back up the way it started.
- `size()` reports `HashTable::allocation_size()`, the real hashbrown allocation including the control bytes, in place of the old estimate.
- A new `clear_and_release` drops every allocation the map holds, and `clear_shrink` calls it.
- The value buffer rounds each growth up to a power of two. A lazily grown buffer and a pre-allocated one then sit on one ladder, so a lazy map is never the larger of the two for the same values. Growth stays geometric.
- `benches/arrow_bytes_map.rs` moves to `with_capacity` so it keeps measuring the pre-allocating constructor.

## What is the testing strategy for this PR?

Two tests in `datafusion/core/tests/memory_limit/mod.rs`, `group_by_count_distinct_utf8` and `group_by_count_distinct_utf8_view`, turn the headline claim into a pass or a fail rather than a number. They run the reproduction query over 4,000 groups with spilling disabled and `target_partitions` pinned to 1, so completing means the query fits the budget rather than spills out of it. The limits are 8 MB and 16 MB, at least 4x clear of both cliffs in the table above. Both tests fail on the merge base and pass here, over five consecutive runs. The `avg(payload)` in the query is load bearing; appendix C says why.

Unit tests cover what the memory-limit tests cannot see: that `new` allocates nothing, that `with_capacity` reports a table size bracketed by an independently derived lower bound, that `take` preserves the configured capacity, that `clear_shrink` drops the reported size to near zero, and that a lazily grown buffer never exceeds a pre-allocated one holding the same values.

Run locally on the rebased head, all passing: `datafusion-physical-expr-common` (87 lib, 8 doc), `datafusion-functions-aggregate-common` (49), `datafusion-functions-aggregate -- count_distinct` (2), `datafusion-physical-plan -- group_values` (96) and the `memory_limit` module (39, which includes the `count_distinct_spill` test that arrived on `main` in #24888 and #24918). `cargo fmt --check` and `cargo clippy --all-targets -D warnings` are clean on the changed crates. CI has not yet run this branch against the new base.

No query results change.

## Are there any user-facing changes?

Yes, in `datafusion-physical-expr-common`. `ArrowBytesMap::new` and `ArrowBytesViewMap::new` no longer pre-allocate, and callers that want the previous behavior should use the new `with_capacity`. Both types also gain `clear_and_release`. This changes an existing public constructor rather than adding one, so tell me if you would like the `api change` label.

For users, a grouped `COUNT(DISTINCT)` on string and binary columns uses much less memory and reports its usage to the `MemoryPool` accurately. A query that previously hit a memory limit may now succeed.

---

## Appendix A: Query 0 costs 6.3% more

Q0 is `COUNT(DISTINCT)` over three high-cardinality strings with no `GROUP BY`. It is a handful of maps that each grow to millions of entries, which is the opposite population from the one this PR targets. The pre-allocation was never the dominant cost there, so removing it buys nothing.

Over six runs the base spans 796.8 to 834.8 MiB and this branch spans 846.5 to 885.5 MiB. The ranges do not overlap, so the effect is real and not run-to-run noise.

Three local builds on a deterministic subset separate the two changes. The middle build differs from the base only in the accounting, because restoring the warm-up makes the constructors byte-identical to base:

| build | Q0 pool peak |
| --- | --- |
| base | 48,421,820 |
| this branch with the warm-up restored | 51,048,396 |
| this branch | 51,018,828 |

That decomposes the increase exactly. +2,626,576 is the accounting correction: `allocation_size()` charges the real hashbrown allocation, which is `4 * buckets + 5,384` more than the old formula, being the control bytes plus the 7/8 load-factor slack. -29,568 is the lazy constructor, which makes Q0 slightly better.

Reverting the accounting would restore an under-report of about 19% on this path. That under-report is the bug this PR exists to fix, and the memory-limit result above depends on fixing it.

## Appendix B: what one accumulator costs

One per-group accumulator holding a single 24-byte value:

| | before, actual | before, reported | after |
| --- | --- | --- | --- |
| `BytesDistinctCountAccumulator` | 14,648 B | 8,240 B | 180 B |
| `BytesViewDistinctCountAccumulator` | 33,920 B | 28,792 B | 260 B |

The middle column is the reporting gap. The `Utf8` map really held 14,648 bytes and reported 8,240, because the whole hash table was invisible to the old accounting.

These are measured directly rather than asserted in a test, since the exact numbers follow the hashbrown layout.

## Appendix C: notes on the tests and the benchmarks

The memory-limit query uses `avg(payload)`, not `count(*)`. A non-distinct `count` lets `SingleDistinctToGroupBy` rewrite the distinct aggregate into a plain two-stage `GROUP BY`. The per-group accumulators would then never exist, and the tests would pass by construction on the base commit too. That rule accepts a non-distinct `sum`, `min` or `max` because each re-aggregates its own partial results correctly over the deduplicated inner group by. `avg` does not, so the rule can never admit it under any extension, including the one #24859 proposes.

The benchmark figures were measured against the previous merge base `da89c7c85b`. They have not been re-run against the current base `20d1c56761`. The commits after `84f07da` on this branch touch only `datafusion/core/tests/memory_limit/mod.rs`, so nothing on this branch since then can move a benchmark, but the base itself has moved.

Pool peak is the instrument here, not peak RSS. Pool peak reproduces to under 1% on a 98 MiB query and exactly on the 3.4 and 11.7 MiB ones. Peak RSS on this harness has a 4.1% standard deviation over 11 readings of identical code plus a 2.3% order bias, and shows no effect from this change once that null is accounted for.

## Follow-ups, not in this PR

- The same undercount remains at five other production `insert_accounted` call sites: `group_values/row.rs:171`, `multi_group_by/mod.rs:434,554`, `multi_group_by/dictionary.rs:197,584` and `array_agg.rs:989`. Each is one map per query, so the absolute error is bounded, and the fix is the same one-line swap.
- The `count_distinct_groups` benchmarks in `datafusion/functions-aggregate/benches/count_distinct.rs` cover `Int64`, `Int32` and `UInt32` only, so this path has no criterion coverage.

### CODE REVIEW

**Comment by adriangb** [comment]
run benchmark clickbench_partitioned external_aggr
env:
  DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G

**Comment by adriangb** [comment]
run benchmark clickbench_extended
env:
  DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G

**Comment by adriangb** [comment]
## Summary of benchmarks and measurements

Consolidating the evidence for this PR in one place, including the parts that are not yet settled.

### 1. Minimum viable memory budget (strongest evidence)

Two tests added in `datafusion/core/tests/memory_limit/`. The query is
`select group_key, count(distinct value), avg(payload) from t group by group_key`
over 4,000 keys with 2 distinct short values each, `target_partitions = 1`, spilling disabled.
The minimum memory limit at which it completes, bisected on each side:

| value column | before | after | reduction |
| --- | --- | --- | --- |
| `Utf8` | fails 34 MB, passes 36 MB | fails 1.8 MB, passes 2.0 MB | ~18x |
| `Utf8View` | fails 120 MB, passes 124 MB | fails 2.4 MB, passes 2.6 MB | ~46x |

The tests are pinned at 8 MB and 16 MB, at least 4x clear of both cliffs, and were verified to
fail on the merge-base and pass on this branch, with five consecutive repeat runs green.

The `avg(payload)` is deliberate rather than incidental: a non-distinct `count` would let
`SingleDistinctToGroupBy` rewrite the aggregate away, so the per-group accumulators would never
exist and the test would pass for the wrong reason. `avg` cannot be admitted by that rule under
any extension, because averaging per-group averages is arithmetically wrong.

### 2. Per-accumulator footprint

One per-group accumulator holding a single 24-byte value, measured on the merge-base and on this
branch:

| | before, actual | before, reported | after |
| --- | --- | --- | --- |
| `BytesDistinctCountAccumulator` | 14,648 B | 8,240 B | 180 B |
| `BytesViewDistinctCountAccumulator` | 33,920 B | 28,792 B | 260 B |

Two separate defects show up in that table. The pre-allocation is the large one. The reporting
gap is the second: `ArrowBytesViewMap` seeded `map_size` from
`capacity() * size_of::<Entry<V>>()`, which omits the control bytes and under-reports by 1.18x,
while `ArrowBytesMap` seeded it with `0` despite pre-allocating, so any map staying under its
pre-allocated capacity reported its hash table as free indefinitely.

### 3. Benchmarks, memory

`clickbench_extended` at `DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G`
([results](https://github.com/apache/datafusion/pull/24857#issuecomment-5500755168)).
Q2 is `SELECT "BrowserCountry", COUNT(DISTINCT "SocialNetwork"), COUNT(DISTINCT "HitColor"),
COUNT(DISTINCT "BrowserLanguage"), COUNT(DISTINCT "SocialAction") FROM hits GROUP BY 1`, which is
the only query in the benchmark suites that puts a grouped `COUNT(DISTINCT)` on a non-integer
column and therefore the only one that reaches this code:

| query | base | changed | change |
| --- | --- | --- | --- |
| Q2 (grouped, 4 string distincts) | 98.6 MiB | 11.7 MiB | **-88.1%** |
| Q1 (ungrouped string distincts) | 3.4 MiB | 2.4 MiB | -29.4% |
| Q12 | 1.3 MiB | 1.0 MiB | -21.2% |

Everything else moves by less than 5% in either direction. Worth stating plainly: this rests on a
single run of a single query, because no other benchmark query exercises the path. The
memory-limit tests in section 1 are the more reliable evidence.

Other suites at the same limit:
[`clickbench_partitioned`](https://github.com/apache/datafusion/pull/24857#issuecomment-5500500275),
[`external_aggr`](https://github.com/apache/datafusion/pull/24857#issuecomment-5500609445).
Neither contains a grouped non-integer distinct count, so neither shows movement, as expected.

### 4. Benchmarks, latency

No effect, which is the expected result for an allocation-sizing change. `clickbench_extended`
totals 37,961 ms against 37,367 ms, with 3 queries faster, 1 slower and 10 unchanged, all inside
run-to-run noise. No query failed and no spilling occurred on either side.

### 5. Open: peak RSS

Peak process RSS rose on all three runs: 9.3 to 9.6 GiB on `clickbench_partitioned`, 418.8 to
445.6 MiB on `external_aggr`, and 9.9 to 10.7 GiB on `clickbench_extended`. The direction is
consistent and the magnitude roughly tracks how much each suite uses the changed path.

It is not yet clear whether that is real. A companion accounting-only change that cannot alter
allocation moved RSS by -8.1%, -7.4% and +3.6% across the same three suites, so this harness shows
at least +/-8% single-run spread and these deltas sit inside it. Repeat runs, including
base-against-base to establish a null distribution, are in progress and will be posted here.

If the effect is real, the likely mechanism is allocator retention rather than live memory:
growing a table from zero by doubling costs log2(N) reallocations per accumulator where there was
previously one up-front allocation, and freed blocks return to allocator free lists rather than to
the OS. The mitigation would be a small non-zero initial capacity, on the order of 8 entries,
which keeps nearly all of the reduction (most per-group sets hold fewer than ten values) while
eliminating the growth steps for the common case. That variant will be measured before any
recommendation is made.

**Comment by adriangb** [comment]
run benchmark clickbench_extended clickbench_extended clickbench_extended

# Noise floor, not a comparison of this PR.
# Both sides are pinned to da89c7c8, the merge-base the earlier
# clickbench_extended run used, so the two sides compile identical code and
# every difference reported below is harness noise. Three jobs give three
# independent A/A readings of peak RSS and of the per-query pool peaks.
env:
  DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G
baseline:
  ref: da89c7c85b1c888ba4038317eacbdcfe03ab9b1a
changed:
  ref: da89c7c85b1c888ba4038317eacbdcfe03ab9b1a

**Comment by adriangb** [comment]
run benchmark clickbench_extended clickbench_extended clickbench_extended

# Three repeats of the comparison in
# https://github.com/apache/datafusion/pull/24857#issuecomment-5500755168,
# with both sides pinned so the repeats measure the same two commits.
# 84f07da is the last commit on this branch that changes benchmarked code;
# the two commits after it touch only
# datafusion/core/tests/memory_limit/mod.rs.
env:
  DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G
baseline:
  ref: da89c7c85b1c888ba4038317eacbdcfe03ab9b1a
changed:
  ref: 84f07dadabe68e25900b59909c9c0c28e99c7c56

**Comment by adriangb** [comment]
## Repeat runs: RSS question resolved, and two corrections

Following up on the open item in the summary above. Six further runs of `clickbench_extended` at
`DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G`, three of them base-against-base so both sides are the
identical commit, which gives a null distribution to measure everything else against.

### Peak RSS: no effect. The earlier increase was a single high draw.

Peak RSS at the merge-base across 11 readings of identical code: 9.4, 9.5, 9.7, 9.9, 10.0, 10.1,
10.2, 10.3, 10.4, 10.6, 10.6 GiB. Mean 10.06 GiB, standard deviation 0.41 GiB (4.1%), and a
max/min span of **12.8%**.

Within-run RSS delta on identical code: 0.0%, -2.9%, -4.0%. On the three base-vs-branch repeats:
**-2.9%, -2.1%, -6.7%**, against the +8.1% originally reported. Welch t about -0.87, p about 0.43.

There is also a mild order effect: the base side runs first in each pod and the second side reads
about 2.3% lower on average, so the original +8.1% was measured against that bias rather than with
it. The allocator-churn hypothesis is not supported and no mitigation is warranted.

### Query 2 reproduces exactly. The single-run caveat is retired.

Null spread on identical code, 6 readings: 98.1, 98.1, 98.2, 98.2, 98.2, 98.6 MiB, a mean delta of
+0.10% with standard deviation 0.27. Pool peak reproduces to well under 1%.

Across the three repeats the changed side read 11.7 MiB every time, giving **-88.12%, -88.09%,
-88.12%**, standard deviation 0.02 percentage points. With the original run that is four for four
at -88.1%.

It is also mechanistically predicted rather than only measured. `Entry<()>` in the view map is a
`u128` plus a `u64`, 32 bytes aligned, and `HashTable::with_capacity(512)` rounds to 1024 buckets,
so each pre-allocated per-group set costs about 33.8 KiB. Q2 is four `COUNT(DISTINCT)` aggregates
grouped by `BrowserCountry`, roughly 2,900 accumulators, predicting about 96 MiB against the 98.3
MiB observed.

Query 1 also holds: 3.4 MiB base and 2.4 MiB changed in every run, with zero variance on either
side.

### Correction 1: the Query 12 number was noise. Retracting it.

The summary above cited Q12 at -21.2%. That does not survive. The null spread on Q12 is -7.1% to
+18.2% (standard deviation 14.6) on identical code, and the repeats gave -9.1%, -9.1% and +7.7%.
Q12 is `MAX(FIRST_VALUE(...) GROUP BY "OS")` with no `COUNT(DISTINCT)` anywhere in it, so there is
no mechanism by which this change could touch it. It is roughly a 1.2 MiB quantity reported to 0.1
MiB granularity. Please disregard that row.

### Correction 2: Query 0's pool peak genuinely increases, by about 4.2%

This is a real, opposite-signed cost that the first run did not surface, and it should be on the
record.

Base commit, 11 readings: 796.8 to 840.8 MiB, mean 821.9. Branch, 4 readings: 846.5, 850.5, 858.5,
871.0 MiB, mean 856.6. **All four branch readings exceed all eleven base readings** (exact rank
p = 0.0007). Restricting to second-position readings only, to control for the order effect, all
four branch values still exceed all three base values: +4.2%, p = 0.029, about +35 MiB.

Q0 is `COUNT(DISTINCT)` over three high-cardinality strings with **no** `GROUP BY`, so a handful of
maps that each grow to millions of entries. That is the opposite population from the one this
change targets, and the result is the expected flip side of removing a warm-up: a map destined to
get large now reaches its size through more doubling steps, and the pool records the transient
where the old and new tables are both live.

The overall picture is coherent. Ungrouped and low cardinality (Q1) improves because the
pre-allocation dominated; grouped with many tiny sets (Q2) improves enormously for the same
reason; ungrouped and high cardinality (Q0) costs a little, because the pre-allocation was never
the dominant term there.

### A note on this harness that applies beyond this PR

Pool peak is a trustworthy instrument here and peak RSS is not. Pool peak reproduced to under 1% on
a 98 MiB query and exactly on 3.4 and 11.7 MiB ones. Peak RSS has a 4.1% standard deviation plus a
2.3% systematic order bias, so any claim resting on a single-run RSS figure needs at least three
runs and a null. Pool figures around 1 MiB, reported to 0.1 MiB, need repeats too, which is what
caught Q12.

Result comments: base-vs-base
[1](https://github.com/apache/datafusion/pull/24857#issuecomment-5513902626),
[2](https://github.com/apache/datafusion/pull/24857#issuecomment-5513918481),
[3](https://github.com/apache/datafusion/pull/24857#issuecomment-5513921246);
base-vs-branch
[1](https://github.com/apache/datafusion/pull/24857#issuecomment-5513878693),
[2](https://github.com/apache/datafusion/pull/24857#issuecomment-5513885404),
[3](https://github.com/apache/datafusion/pull/24857#issuecomment-5513887805).

**Comment by adriangb** [comment]
run benchmark clickbench_extended clickbench_extended clickbench_extended
env:
  DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G

**Comment by adriangb** [comment]
run benchmark clickbench_extended clickbench_extended
env:
  DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G

**Comment by adriangb** [comment]
run benchmark clickbench_extended
env:
  DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G

**Comment by adriangb** [comment]
## Query 0: six runs, and a correction to the mechanism I gave earlier

An earlier comment on this PR put Query 0's pool-peak increase at +4.2% from four readings, and
explained it as a transient where an old and a new hash table are both live during a doubling.
**That mechanism is wrong.** Both the figure and the explanation are corrected below.

### Six runs on the current head

All six ran `clickbench_extended` at `DATAFUSION_RUNTIME_MEMORY_LIMIT: 4G` against head
`b9e6252ece`.

| run | base | changed | change |
| --- | --- | --- | --- |
| 1 | 814.8 MiB | 877.0 MiB | +7.6% |
| 2 | 834.8 MiB | 885.5 MiB | +6.1% |
| 3 | 812.8 MiB | 871.0 MiB | +7.2% |
| 4 | 826.8 MiB | 873.0 MiB | +5.6% |
| 5 | 814.8 MiB | 846.5 MiB | +3.9% |
| 6 | 796.8 MiB | 856.5 MiB | +7.5% |

Mean +6.3%. The base readings span 796.8 to 834.8 MiB and the changed readings span 846.5 to
885.5 MiB, so the two ranges do not overlap. This is a real effect, not run-to-run noise.

### It is the accounting correction, not new allocation

Measured locally on a deterministic 3-file subset with `-n 1`, three builds of the same data and
command. The middle build differs from the base **only** in the accounting fix, because restoring
`with_capacity(128)` plus the 8 KiB buffer makes the constructors byte-identical to base:

| build | Q0 pool peak |
| --- | --- |
| base | 48,421,820 |
| branch with the warm-up restored | 51,048,396 |
| branch head | 51,018,828 |

That decomposes the increase exactly:

- **+2,626,576 is the accounting correction.** `map_size` charged
  `capacity * size_of::<Entry>()` and only on growth. `allocation_size()` charges the real
  hashbrown allocation, which is `4 * buckets + 5,384` more: the control bytes plus the 7/8
  load-factor slack. At full scale `SearchPhrase` holds about 6.02M distinct values, so 2^23
  buckets, which predicts about 32.0 MiB and matches the measured band.
- **−29,568 is the lazy constructor.** It makes Q0 slightly better, not worse.

The transient explanation fails on two counts. The pool samples `size()` only between batches, so
a transient inside `insert_unique` is invisible to it. And the base performed the same doublings
above 224 entries anyway.

Query 0 is `COUNT(DISTINCT)` over three high-cardinality strings with **no** `GROUP BY`: a handful
of maps that each grow to millions of entries. That is the opposite population from the one this
PR targets, so the pre-allocation was never the dominant cost there and removing it buys nothing.

### Why this is not a regression to fix

The extra 6.3% is memory the query always held and the pool could not see. Reverting the change
restores an under-report of about 19% on this path. That under-report is the bug this PR exists to
fix, and it is what the memory-limit improvement depends on: the minimum viable budget for a
grouped `COUNT(DISTINCT Utf8)` goes from 36 MB to 2.0 MB, and for `Utf8View` from 124 MB to 2.6 MB.

Restoring the warm-up was tested and makes Query 0 **worse**, so the "capacity hint from
`GroupsAccumulatorAdapter`" follow-up listed earlier is withdrawn. It is measurably
counterproductive.

### Query 2 across the same six runs

The changed side reported **11.7 MiB in every run**, against a base of 98.1 to 98.6 MiB. Six
consecutive readings of -88.1% with no variance on the changed side, and nine measurements in
total including the earlier runs.

**Comment by adriangb** [comment]
@2010YOUY01 @kosiew would either of you be interested in reviewing this change?

File: datafusion/physical-expr-common/src/binary_map.rs [hunk 1]
```diff
@@ -663,6 +749,155 @@ mod tests {
...
+
+        // The configured capacities survive, so the map warms back up when it
+        // is emitted from again.
+        map.take();
+        assert!(map.map.capacity() >= INITIAL_MAP_CAPACITY);
+        assert_eq!(map.buffer.capacity(), INITIAL_BUFFER_CAPACITY);
+    }
+
+    #[test]
+    fn lazy_and_pre_allocated_buffers_grow_on_the_same_ladder() {
```
Comment (kosiew) [inline datafusion/physical-expr-common/src/binary_map.rs:937]: Could we add a small boundary test for `push_value_bytes`? It would be useful to cover an empty append, filling exactly to a power-of-two capacity, appending one byte beyond it, and a single append larger than the initial 8192-byte buffer. I'd check that contents are preserved and that the buffer does not grow when its existing capacity is sufficient.

The convergence test already gives good higher-level coverage, but it compares capacities after 1000-value batches. A focused test here would pin the helper's boundary behavior directly, independently of both maps using the same helper.

This is just non-blocking test hardening, not a correctness issue I've found.

**Review by kosiew — APPROVED** [review]
@adriangb, thanks for working on this. The split between lazy allocation for the per-group distinct-count sets and pre-allocation for the long-lived GROUP BY maps makes sense to me. The updated hashbrown accounting and releasing allocations through clear_shrink also look good.

I left one small, non-blocking suggestion to add direct boundary coverage for the new byte-buffer growth helper.

File: datafusion/physical-expr-common/src/binary_map.rs (hunk 1 above)
Comment (adriangb) [reply datafusion/physical-expr-common/src/binary_map.rs:937]: Thanks! Addressed in 7a1a4df.

**Comment by kosiew** [comment]
@adriangb 
re ungrouped_utf8_view_accumulator_is_never_worse_than_a_pre_allocated_set, 
could we split this test matrix based on the hash mode?

The 100,000 and 500,000 cardinality cases also run in the `hash collisions (amd64)` CI job, which enables `datafusion-common/force_hash_collisions`. In that configuration every distinct value hashes to the same bucket, so each insertion has to search an increasingly long collision chain. Since both the lazy and pre-allocated sets receive every value, these larger cases become quadratic and can make the job run for hours or hit its 360-minute timeout, as seen in #25011.

I suggest keeping a small, bounded case under forced collisions to verify collision correctness, while leaving the large cardinalities for the normal hash configuration where they are useful for checking the allocation and capacity-convergence invariant. It would also be good to run the bounded collision case with the same CI profile and forwarded feature before merging.
